org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,244 @@
+                }
+            }
+        }
+    }
+
+    private void cleanUpIndicesIfExists(DfsPackKey key) {
+        if (key != null) {
+            DfsPackDescription description = reversePackDescriptionIndex.remove(key);
+            if (description != null) {
+                packFileCache.remove(description);
+            }
+        }
+    }
+
+    boolean shouldCopyThroughCache(long length) {
+        return length <= maxStreamThroughCache;
+    }
+
+    DfsPackFile getOrCreate(DfsPackDescription description, DfsPackKey key) {
+        return packFileCache.compute(description, (DfsPackDescription k, DfsPackFile v) -> {
+            if (v != null && !v.invalid()) {
+                return v;
+            }
+            DfsPackKey newPackKey = key != null ? key : new DfsPackKey();
+            reversePackDescriptionIndex.put(newPackKey, description);
+            return new DfsPackFile(this, description, newPackKey);
+        });
+    }
+
+    int getBlockSize() {
+        return blockSize;
+    }
+
+    // do something when the block is invalid
+    DfsBlock getOrLoad(DfsPackFile pack, long position, DfsReader dfsReader) throws IOException {
+        final long requestedPosition = position;
+        position = pack.alignToBlock(position);
+        DfsPackKey key = pack.key;
+
+        Ref<DfsBlock> loadedBlockRef = dfsBlockAndIndicesCache.get(new DfsPackKeyWithPosition(key, position), keyWithPosition -> {
+            try {
+                DfsBlock loadedBlock = pack.readOneBlock(keyWithPosition.getPosition(), dfsReader);
+                key.cachedSize.getAndAdd(loadedBlock.size());
+                return new Ref(keyWithPosition.getDfsPackKey(), keyWithPosition.getPosition(),
+                        loadedBlock.size(), loadedBlock);
+            } catch (IOException e) {
+                return null;
+            }
+        });
+
+        if (loadedBlockRef != null) {
+            DfsBlock loadedBlock = loadedBlockRef.get();
+            if (loadedBlock != null && loadedBlock.contains(key, position)) {
+                return loadedBlock;
+            }
+        }
+
+        // the current block is not valid, remove it and attempt `getOrLoad` again
+        dfsBlockAndIndicesCache.invalidate(new DfsPackKeyWithPosition(key, position));
+        return getOrLoad(pack, requestedPosition, dfsReader);
+    }
+
+    void put(DfsBlock block) {
+        put(block.pack, block.start, block.size(), block);
+    }
+
+    public <T> Ref<T> put(DfsPackKey key, long position, int size, T value) {
+        return dfsBlockAndIndicesCache.get(new DfsPackKeyWithPosition(key, position), keyWithPosition -> {
+            // if it's not an index, increase the cached size
+            if (keyWithPosition.position >= 0) {
+                keyWithPosition.getDfsPackKey().cachedSize.getAndAdd(size);
+            }
+            return new Ref(keyWithPosition.getDfsPackKey(), keyWithPosition.getPosition(), size, value);
```

Comment:
Do we have guarantees that this method is only called if this wasn't present in the map?  I'm worried about something getting put in the map twice and cachedSize getting incremented twice (same question with getOrLoad above)